Fix image optimization in shortcodes within content blobs

felixarntz opened the pull request:
* Move `wp_filter_content_tags` to a later priority than `do_shortcode()` is at (>11, for now 12 was chosen)
* Bail early when encountering a shortcode that appears within content blobs like `the_content`, `widget_text_content`, `widget_block_content` so that images within those shortcodes are only handled once as part of the overall content blob
* Add test coverage for a shortcode within post content, as well as a shortcode within a block within post content
* Remove test cases which aren't relevant anymore given that shortcodes should be treated within their content blobs rather than individually

Trac ticket: https://core.trac.wordpress.org/ticket/58853

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Commit:
Fix image optimization in shortcodes within content blobs.

Commit:
Correct inline comment.

Commit:
Merge branch 'trunk' into fix/58853-shortcode-image-optimization

Commit:
Fix tests.

Commit:
Merge branch 'trunk' into fix/58853-shortcode-image-optimization

felixarntz commented:
@joemcgill Friendly ping, could you please take a look at this one? :)

joemcgill commented on src/wp-includes/default-filters.php:
```diff
@@ -195,15 +195,15 @@
 add_filter( 'the_content', 'shortcode_unautop' );
 add_filter( 'the_content', 'prepend_attachment' );
-add_filter( 'the_content', 'wp_filter_content_tags' );
+add_filter( 'the_content', 'wp_filter_content_tags', 12 ); // Runs after do_shortcode().
```
Changing the priority here also has the effect of requiring anyone wanting to remove this filter to update their code to add the new priority as a third param to `remove_filter()`. That said, [a quick search of the WP Directory](https://wpdirectory.net/search/01H9P3ZF7V9W0VDYWGJESSBFNX) doesn't show any plugins that are actually trying to remove this filter, so the change should be safe to make.

As a nitpick, I'd suggest moving these lines to the end of each group so that adding this filter comes after adding the `do_shortcode` filter, or to the end of each group where `do_shortcode` does not apply. This will help with readability.

felixarntz replied:
Regarding the priority change, also see https://core.trac.wordpress.org/ticket/58853#comment:2 for the research I did prior to the PR.

I've updated the order for better readability in https://github.com/WordPress/wordpress-develop/pull/4973/commits/82dc4f3f1c501c6540cd7d7a4341b8261c4a5d15. I've also colocated the `the_content` filter for `do_shortcode` with all the other `the_content` ones, whereas before it was completely separated without a good reason (it had a "shortcodes" comment above, but all other `do_shortcode` callbacks were not there, so that really didn't make much sense anymore).

joemcgill commented on tests/phpunit/tests/media.php:
```diff
@@ -4664,6 +4664,117 @@ public function test_the_excerpt_does_not_affect_omit_lazy_loading_logic() {
+		remove_shortcode( 'full_image' );
+		unregister_block_type( 'core/full-image-shortcode' );
+
+		$this->assertSame( $expected_content, $content );
```
While this should help us catch any regressions here, the way this builds the expected content and runs the assertion isn't as explicit as it could be. Is it possible to adjust this test so the assertions are focused on each specific thing we're trying to test, i.e.:

1. the first image has the correct `fetchpriority` attribute
2. the second two images are not lazy loaded
3. the final image does have the correct `loading` attribute

Alternately, if this test is meant to only focus on the shortcode issue described in [#58853](https://core.trac.wordpress.org/ticket/58853), can it be written in a way that makes that clear?

felixarntz replied:
@joemcgill Happy to add more specific tests for those points, though I think a comprehensive test like this is also useful, kind of similar to how a mix of integration and unit tests makes for the best overall coverage. I'll add a few more specific tests.

joemcgill replied:
Thanks. My main concern here is that it's clear to future developers (including future us) what exactly is being tested (i.e., the intent of the test) and also that it is clear if the assertion(s) fail, why the assertion has failed.

felixarntz replied:
@joemcgill I have added one more test which is a bit simpler covering only a single image shortcode together with a hard-coded image. It's still rather complex though. What I have therefore done is add more comprehensive inline documentation to both of these tests, to explain what is going on.

Unfortunately there are so many intertwined considerations at play so that, in order to cover the functionality, the tests end up quite complex. For example, using a single shortcode in a test wouldn't be able to show much. And we always need some of the more complex logic such as use a main query loop etc to simulate the requirements that `wp_get_loading_optimization_attributes()` relies on to make decisions.

Please take another look. While it's still complex, I hope with the additional comments and one at least slightly simpler test for what is less of an edge-case it is more clear now.

costdev replied:
@felixarntz Would `$this->assertStringContainsString( $expected_image_markup, $content, 'The first image should...' );` work for you? That way, you're essentially (probably not literally though) saving each `$replace` part of `str_replace()` to a variable to be tested in its own assertion with a message adding context.

felixarntz replied:
@costdev I don't think that would work, since this test should assert that all images have the expected markup, plus their order in the string matters too, hence a full comparison is most reliable.

joemcgill replied:
Thanks @felixarntz, this is better, but rather than building a single `$expected_content` blob, I do think writing an assertion for each specific image—as @costdev suggests—to make sure it includes the expected optimization attributes would be clearer if/when things fail. If you need an additional assertion to make sure all images show up in the markup in the correct order together (though, I don't think that's all that critical) then you could include a test for that in addition to the individual assertions.

joemcgill commented on tests/phpunit/tests/media.php:
```diff
@@ -5007,63 +5118,17 @@ public function test_wp_get_loading_optimization_attributes_low_res_image() {
-	 * @dataProvider data_wp_get_loading_optimization_attributes_in_shortcodes
 	 */
-	public function test_wp_get_loading_optimization_attributes_in_shortcodes( $setup, $expected, $message ) {
+	public function test_wp_get_loading_optimization_attributes_in_shortcodes() {
```
Rather than removing all of the different use cases that were previously being tested, can we update the data provider to ensure the `do_shortcode` context is performing as expected given the logic changes made in the switch statement in `wp_get_loading_optimization_attributes`?

felixarntz replied:
@joemcgill The updated test is effectively the same as the third scenario from the original data provider which IMO is the only scenario still relevant with the new logic. The main query loop consideration, which was what the other two scenarios were covering is no longer relevant now.

The one thing that replaces that logic for which right now there's no concrete test is to cover the early return when using the `do_shortcode` context within e.g. `the_content` filter. I'll add tests for that.

joemcgill replied:
Right. It makes sense that the scenarios need to be updated, since some of the originals no longer apply. I just want to make sure we have coverage for all expected use cases.

felixarntz replied:
@joemcgill I have added another test with a data provider to cover the `wp_get_loading_optimization_attributes()` changes for early return related to this specifically.

joemcgill reviewed (commented):
Without doing a lot of manual testing, this looks like it should fix the issues we've identified. I've left a few suggestions, and would like to see if we could make the unit tests a bit more explicit if possible.

Commit:
Merge branch 'trunk' into fix/58853-shortcode-image-optimization

Commit:
Update do_shortcode and wp_filter_content_tags filter order in default-filters.php for better readability.

Commit:
Add further test coverage and add more inline docs for complicated tests.

spacedmonkey commented on src/wp-includes/default-filters.php:
```diff
@@ -195,16 +195,17 @@
-add_filter( 'the_content', 'wp_filter_content_tags' );
 add_filter( 'the_content', 'wp_replace_insecure_home_url' );
+add_filter( 'the_content', 'do_shortcode', 11 ); // AFTER wpautop().
+add_filter( 'the_content', 'wp_filter_content_tags', 12 ); // Runs after do_shortcode().
```
If a developer have unhooked on priority 10 if they don't want image features. Looking online, I can't see anyone doing this, but it worth noting.

felixarntz replied:
Thanks @spacedmonkey, see https://github.com/WordPress/wordpress-develop/pull/4973#discussion_r1320204338. I did a bunch of research already, and this change should be safe to make. Needless to say there will be a dev note, but it doesn't look like this is going to be a big problem since there is zero usage in the plugin repository.

spacedmonkey committed:
Merge branch 'trunk' into fix/58853-shortcode-image-optimization

felixarntz commented:
Noting here that this PR will need to be refreshed if #5197 is committed first, or vice-versa if this one is committed first. cc @mukeshpanchal27

Commit:
Merge branch 'trunk' into fix/58853-shortcode-image-optimization

Commit:
Merge branch 'fix/58853-shortcode-image-optimization' of github.com:felixarntz/wordpress-develop into fix/58853-shortcode-image-optimization

felixarntz commented:
@joemcgill @spacedmonkey Would you mind giving this another review please? I have addressed your feedback and refreshed the PR now that https://core.trac.wordpress.org/changeset/56612 has been committed.

costdev commented on tests/phpunit/tests/media.php:
```diff
@@ -5173,64 +5405,67 @@ public function test_wp_get_loading_optimization_attributes_low_res_image() {
+	 * @ticket 58853
+	 * @dataProvider data_get_filters_with_do_shortcode_callback
+	 *
+	 * @covers ::wp_get_loading_optimization_attributes
```
```suggestion
	 * @covers ::wp_get_loading_optimization_attributes
	 *
	 * @param string $filter_name The name of the filter to hook.
```

felixarntz replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/4973/commits/79ebdddfceec96dec01e20c0bcad8be8fb653b82

costdev commented on tests/phpunit/tests/media.php:
```diff
@@ -5173,64 +5405,67 @@ public function test_wp_get_loading_optimization_attributes_low_res_image() {
+	 * @return array[]
+	 */
+	public function data_get_filters_with_do_shortcode_callback() {
+		return array(
```
We can use this helper method to make this use named datasets while also making the datasets easier to read/add/modify.

```php
return self::text_array_to_dataprovider(
	array(
		'the_content',
		'widget_text_content',
		'widget_block_content',
	)
);
```

Produces this before passing to the test method:
```php
array(
	'the_content'          => array( 'the_content' ),
	'widget_text_content'  => array( 'widget_text_content' ),
	'widget_block_content' => array( 'widget_block_content' ),
)
```

felixarntz replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/4973/commits/79ebdddfceec96dec01e20c0bcad8be8fb653b82

Commit:
Merge branch 'trunk' into fix/58853-shortcode-image-optimization

Commit:
Address test feedback.

joemcgill reviewed (approved):
I finished another pass at a review and did some extensive testing of this with a few use cases locally to make sure the complicated logic was being executed in the order I expected and this seems ready to go to me. Nice work @felixarntz.

I still think we could improve the PHPUnit coverage a bit to be more specific, but it does cover everything we need at present, so no need to hold up an initial commit.

For my own interest, I processed the same content pattern used in `test_wp_filter_content_tags_handles_shortcode_images_also_in_blocks_within_the_content` locally on both the Twenty Twenty-Three and Twenty Twenty-One themes and the expected result is achieved in both cases.

Interestingly, `wp_get_loading_optimization_attributes()` processes the image from the custom block (which uses `do_shortcode()` in its render callback) before the actual gallery shortcode, but neither are modified since the `do_shortcode` context is passed during the `the_content` hook, so it has no effect on the images before they're processed in content. 

This was one of my original concerns, so it's nice to see that it's a non-issue.

mukeshpanchal27 commented on tests/phpunit/tests/media.php:
```diff
@@ -5173,64 +5405,71 @@ public function test_wp_get_loading_optimization_attributes_low_res_image() {
+	 * Tests that the `do_shortcode` context does not result in loading optimization changes when used within a content
+	 * blob.
+	 *
+	 * @ticket 58853
+	 * @dataProvider data_get_filters_with_do_shortcode_callback
+	 *
+	 * @covers ::wp_get_loading_optimization_attributes
```
```suggestion
	 * @ticket 58853
	 *
	 * @covers ::wp_get_loading_optimization_attributes
	 *
	 * @dataProvider data_get_filters_with_do_shortcode_callback
```

Update annotation position per [56070](https://core.trac.wordpress.org/ticket/56070) proposal.

mukeshpanchal27 commented on tests/phpunit/tests/media.php:
```diff
@@ -4831,6 +4831,238 @@ public function test_the_excerpt_does_not_affect_omit_lazy_loading_logic() {
+		// Add shortcode that prints a large image, and a block type that wraps it.
+		add_shortcode(
+			'full_image',
+			static function( $atts ) {
```
```suggestion
			static function ( $atts ) {
```

mukeshpanchal27 commented on tests/phpunit/tests/media.php:
```diff
@@ -4831,6 +4831,238 @@ public function test_the_excerpt_does_not_affect_omit_lazy_loading_logic() {
+		// Add shortcode that prints a large image, and a block type that wraps it.
+		add_shortcode(
+			'full_image',
+			static function( $atts ) {
```
```suggestion
			static function ( $atts ) {
```

mukeshpanchal27 commented on tests/phpunit/tests/media.php:
```diff
@@ -4831,6 +4831,238 @@ public function test_the_excerpt_does_not_affect_omit_lazy_loading_logic() {
+		register_block_type(
+			'core/full-image-shortcode',
+			array(
+				'render_callback' => static function( $atts ) {
```
```suggestion
				'render_callback' => static function ( $atts ) {
```

mukeshpanchal27 commented on tests/phpunit/tests/media.php:
```diff
@@ -4831,6 +4831,238 @@ public function test_the_excerpt_does_not_affect_omit_lazy_loading_logic() {
+				array( '', '<img fetchpriority="high" ' ),
+				preg_replace_callback(
+					'/gallery-(\d+)/',
+					static function( $match ) {
```
```suggestion
					static function ( $match ) {
```

mukeshpanchal27 commented on tests/phpunit/tests/media.php:
```diff
@@ -5173,64 +5405,71 @@ public function test_wp_get_loading_optimization_attributes_low_res_image() {
+		remove_all_filters( $filter_name );
+		add_filter(
+			$filter_name,
+			function( $content ) use ( &$result ) {
```
```suggestion
			function ( $content ) use ( &$result ) {
```

mukeshpanchal27 reviewed (commented):
Per [Changeset 56559](https://core.trac.wordpress.org/changeset/56559), Include one space after `function` keyword for closures.

spacedmonkey committed:
Merge branch 'trunk' into fix/58853-shortcode-image-optimization

felixarntz commented on src/wp-includes/default-filters.php:
```diff
@@ -195,16 +195,17 @@
-add_filter( 'the_content', 'wp_filter_content_tags' );
 add_filter( 'the_content', 'wp_replace_insecure_home_url' );
+add_filter( 'the_content', 'do_shortcode', 11 ); // AFTER wpautop().
+add_filter( 'the_content', 'wp_filter_content_tags', 12 ); // Runs after do_shortcode().
```
```suggestion
add_filter( 'the_content', 'wp_filter_content_tags', PHP_INT_MAX ); // Runs after any other hooks that may inject images.
```

costdev replied:
It seems [some plugins](https://wpdirectory.net/search/01HB6M3FT27M1BSX4QQV25MD8R) also use `PHP_INT_MAX`. Is this likely to cause a problem?

felixarntz commented on src/wp-includes/default-filters.php:
```diff
@@ -195,16 +195,17 @@
 add_filter( 'the_excerpt', 'shortcode_unautop' );
-add_filter( 'the_excerpt', 'wp_filter_content_tags' );
 add_filter( 'the_excerpt', 'wp_replace_insecure_home_url' );
+add_filter( 'the_excerpt', 'wp_filter_content_tags', 12 );
```
```suggestion
add_filter( 'the_excerpt', 'wp_filter_content_tags', PHP_INT_MAX );
```

felixarntz commented on src/wp-includes/default-filters.php:
```diff
@@ -230,13 +231,13 @@
-add_filter( 'widget_text_content', 'wp_filter_content_tags' );
 add_filter( 'widget_text_content', 'wp_replace_insecure_home_url' );
 add_filter( 'widget_text_content', 'do_shortcode', 11 ); // Runs after wpautop(); note that $post global will be null when shortcodes run.
+add_filter( 'widget_text_content', 'wp_filter_content_tags', 12 ); // Runs after do_shortcode().
```
```suggestion
add_filter( 'widget_text_content', 'wp_filter_content_tags', PHP_INT_MAX ); // Runs after any other hooks that may inject images.
```

felixarntz commented on src/wp-includes/default-filters.php:
```diff
@@ -230,13 +231,13 @@
 add_filter( 'widget_block_content', 'do_blocks', 9 );
-add_filter( 'widget_block_content', 'wp_filter_content_tags' );
 add_filter( 'widget_block_content', 'do_shortcode', 11 );
+add_filter( 'widget_block_content', 'wp_filter_content_tags', 12 ); // Runs after do_shortcode().
```
```suggestion
add_filter( 'widget_block_content', 'wp_filter_content_tags', PHP_INT_MAX ); // Runs after any other hooks that may inject images.
```

felixarntz commented on src/wp-includes/formatting.php:
```diff
@@ -3980,7 +3980,7 @@ function wp_trim_excerpt( $text = '', $post = null ) {
 		 * is wasteful and can lead to bugs in the image counting logic.
 		 */
-		$filter_image_removed = remove_filter( 'the_content', 'wp_filter_content_tags' );
+		$filter_image_removed = remove_filter( 'the_content', 'wp_filter_content_tags', 12 );
```
```suggestion
		$filter_image_removed = remove_filter( 'the_content', 'wp_filter_content_tags', PHP_INT_MAX );
```

felixarntz commented on src/wp-includes/formatting.php:
```diff
@@ -4003,7 +4003,7 @@ function wp_trim_excerpt( $text = '', $post = null ) {
 		 */
 		if ( $filter_image_removed ) {
-			add_filter( 'the_content', 'wp_filter_content_tags' );
+			add_filter( 'the_content', 'wp_filter_content_tags', 12 );
```
```suggestion
			add_filter( 'the_content', 'wp_filter_content_tags', PHP_INT_MAX );
```

felixarntz commented on tests/phpunit/tests/formatting/wpTrimExcerpt.php:
```diff
@@ -129,7 +129,7 @@ public function test_wp_trim_excerpt_should_not_permanently_unhook_wp_filter_con
 		wp_trim_excerpt( '', $post );
 
-		$this->assertSame( 10, has_filter( 'the_content', 'wp_filter_content_tags' ), 'wp_filter_content_tags() was not restored in wp_trim_excerpt()' );
+		$this->assertSame( 12, has_filter( 'the_content', 'wp_filter_content_tags' ), 'wp_filter_content_tags() was not restored in wp_trim_excerpt()' );
```
```suggestion
		$this->assertSame( PHP_INT_MAX, has_filter( 'the_content', 'wp_filter_content_tags' ), 'wp_filter_content_tags() was not restored in wp_trim_excerpt()' );
```

felixarntz commented on tests/phpunit/tests/formatting/wpTrimExcerpt.php:
```diff
@@ -141,7 +141,7 @@ public function test_wp_trim_excerpt_does_not_restore_wp_filter_content_tags_if_
 
 		// Remove wp_filter_content_tags() from 'the_content' filter generally.
-		remove_filter( 'the_content', 'wp_filter_content_tags' );
+		remove_filter( 'the_content', 'wp_filter_content_tags', 12 );
```
```suggestion
		remove_filter( 'the_content', 'wp_filter_content_tags', PHP_INT_MAX );
```

felixarntz reviewed (commented):
Per https://wordpress.slack.com/archives/C02KGN5K076/p1695659619574269, putting this here for consideration:
* The PR changes the filter priority of `wp_filter_content_tags()` to `12`, so that it runs after parsing shortcodes.
* That change itself, while technically a BC break, has been determined to be safe given no relevant usage was found in any plugins, and the function isn't particularly old either.
* Now the question is: Is `12` the priority we want to go with, or should we choose a much higher value (later priority), to also cover images injected by various plugin mechanisms more reliably? One may argue that a very high value, possibly even `PHP_INT_MAX`, would be most appropriate so that generally programmatically injected images are being caught. It would be particularly helpful since `wp_get_loading_optimization_attributes()` intentionally doesn't apply any changes to contexts run as part of e.g. `the_content`, so that images in those contexts are all considered together.

felixarntz commented:
Per the discussion in https://wordpress.slack.com/archives/C02KGN5K076/p1695659619574269, it's safer to keep the priority change to `12` rather than using a later priority.

felixarntz committed:
Update code format based on new requirements.

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

Commit:
Merge branch 'trunk' into fix/58853-shortcode-image-optimization

Commit:
Fix tests after integration of decoding async into wp_get_loading_optimization_attributes().

Commit:
Merge branch 'trunk' into fix/58853-shortcode-image-optimization

felixarntz commented:
Committed in https://core.trac.wordpress.org/changeset/56693

westonruter commented on src/wp-includes/default-filters.php:
```diff
@@ -195,16 +195,17 @@
 add_filter( 'the_content', 'wpautop' );
 add_filter( 'the_content', 'shortcode_unautop' );
 add_filter( 'the_content', 'prepend_attachment' );
-add_filter( 'the_content', 'wp_filter_content_tags' );
```
You have have seen this, but there is one instance of a plugin removing this filter at priority 10:

https://wpdirectory.net/search/01HBEBXA89CXBWDPZHN3C6FWNQ

Nevertheless, the code is commented out 😄 

```php
            /*
             * Temporarily unhook wp_filter_content_tags() since any tags
             * within the excerpt are stripped out. Modifying the tags here
             * is wasteful and can lead to bugs in the image counting logic.
             */
            //$filter_removed = remove_filter( 'the_content', 'wp_filter_content_tags' );
```

westonruter replied:
In a GitHub search, there are ~13 non-core occurrences of this removal: https://github.com/search?q=%2Fremove_filter%5C%28%5Cs*%5B%27%22%5Dthe_content%5B%27%22%5D%2C%5Cs*%5B%27%22%5Dwp_filter_content_tags%5B%27%22%5D%2F&type=code&p=1

westonruter replied:
So overall pretty safe to change the priority. I presume the dev note will call out the priority change.

westonruter replied:
Also, I couldn't think of any specific reason to go with a priority higher than 12. As I found, there are scant removals of this filter in the plugin ecosystem, so if the priority is changed later it's not going to be the end of the world. 

I did do a search for plugins that are filtering `the_content` to do _something_ with images: https://wpdirectory.net/search/01HBHY8RKC5AARSB5J4M4FYJ3H

I didn't find any examples of a plugin adding an image with a priority above 12.

westonruter replied:
All this to say, this will correctly fix the case I made a gist plugin for: https://gist.github.com/westonruter/faccfac310645d0c922f1c56d9ec1fbf
